rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+there is just as much complexity. When we `look_for_files` and build watchers
+for them, we need to make a decision about where to start reading in that file.
+That decision should be based on any stored checkpoint, file metadata (e.g.
+mtime), whether the file was found at startup or runtime, and how the source is
+configured.
+
+Since this decision really only happens in one place, the challenge is more
+about providing an understandable config UI than designing the right interface.
+This should be driven by real world use cases. For example:
+
+1. Ignoring existing checkpoints
+1. Start at the beginning or end of existing files, optionally taking into
+   account factors like mtime
+1. Start at the beginning or end of files added while we're watching (this can
+   be tricky)
+1. Ordering which of the above concerns take precedence
+
+I would suggest a config like the following:
+
+* `ignore_checkpoints = true|false` (still write but don't read existing checkpoints)
+* `read_from = beginning|end` (where to start if there's no checkpoint or
+    they're being ignored)
+* `skip_older_than = duration` (relevant when `read_from = beginning`, seek to
+    end based on mtime)
+* Always start at the beginning of files added while we're watching (it's hard
+    to tell a `mv` from a create and write, so don't rely on seeing an empty file
+    first to get all the data)
+
+If we adopt this, we can also implement a solution where we don't hold open file
+handles for files that match `skip_older_than`, since this has caused some
+issues. Naively, we need the open handle to attempt reads in case new data is
+written to the file, but we could also implement that as a new state of
+`FileWatcher` that stashes the inital size and periodically checks the file
+metadata (mtime and size) to see if it should start reading at what was the end.
+
+### General tweaks
+
+In addition to extracting these various subcomponents, there are some other
+relatively simple changes we can make to help address the issues we discussed.
+
+#### File discovery and checkpointing
+
+We currently use the outdated `glob_minimum_cooldown` config option to determine
+how often to do both of these tasks. We should switch them to their own
+independent config options and allow them to be disabled (e.g. a batch use case
+does not need to continuously look for new files).
+
+We should also move them into their own periodic tasks, outside of the main read
+loop. This should give us better performance and help avoid bad behavior in
+situations where either is expensive (e.g. discovering millions of files on
+EBS).
+
+#### Read concurrency
+
+With most of the other concerns pulled out, we can afford to adjust the main
+read loop to allow for some new capabilities. The most interesting is concurrent
+reads, but it will require a bit of experimentation before we're able to
+determine if it's worthwhile. There are a few possible approaches:
+
+1. Dispatch reads to an explicit threadpool
```

Comment:
It seems to me that the file source is still using sync? Perhaps we could gain some benefits in structure/perf using our tokio runtime and it's async filesystem calls?

Reply:
If/when we change this, that would be my preference as well. I'll make a note of that.